src/CoreWCF.Primitives/src/CoreWCF/Security/Authorization/AuthorizationServiceBehavior.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Reflection;
+using CoreWCF.Channels;
+using CoreWCF.Description;
+using Microsoft.AspNetCore.Authorization;
+
+namespace CoreWCF.Security.Authorization
+{
+    public class AuthorizationServiceBehavior : IServiceBehavior
+    {
+        private readonly IAuthorizationPolicyProvider _authorizationPolicyProvider;
+        private readonly IAuthorizationService _authorizationService;
+
+        public AuthorizationServiceBehavior(IAuthorizationPolicyProvider authorizationPolicyProvider,
+            IAuthorizationService authorizationService)
+        {
+            _authorizationPolicyProvider = authorizationPolicyProvider;
+            _authorizationService = authorizationService;
```

Comment:
I think we should store the IAuthorizationPolicyProvider on DispatchRuntime as that's where ServiceAuthenticationManager and ServiceAuthorizationManager live. DispatchRuntime has a RequiresAuthorization bool which controls whether an AuthorizationBehavior is returned from AuthorizationBehavior.TryCreate. The AuthorizationBehavior would store the IAuthorizationPolicyProvider as well as an IAuthorizationService. We can either add a method called AuthorizePolicyAsync which takes the principal and the AuthorizationPolicy as parameters and gets called in the same place as you are currently inserting the logic, or we could add parameters to the existing Authorize method (and make it async) so the one method can make decisions about both mechanisms. The existing Authorize code throws a FaultException with the the relevant namespace and faultcode etc which the client will understand and throw SecurityAccessDeniedException.

Reply:
I figured that I no longer need the `IAuthorizationPolicyProvider` when handling requests. Instead i'm using it in `TypeLoader` when building `OperationDescription`.
The `IAuthorizationService` is populated into `AuthorizationBehavior` using `ServiceAuthorizationManager` then `DispatchRuntime` and finally `AuthorizationBehavior.TryCreate`.